Refactor zone renderer

bubblobill opened the pull request:
Progress towards resolving #3691 Refactoring ZoneRenderer

### Description of the Change
Moved some constants to ZoneRendererConstants
Split off grid rendering and grid adjustment methods to new class GridRenderer
Split off halo rendering to new class HaloRenderer

### Possible Drawbacks
none

### Documentation Notes
More progress on modularising zone renderer with two new rendering classes for grids and halos.

### Release Notes
N/A

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/4447)
<!-- Reviewable:end -->

bubblobill committed:
renamed utility method getPoint to getPointAtVector

bubblobill committed:
renamed utility method getPoint to getPointAtVector
renamed utility method lighter to lighten

bubblobill committed:
Merge remote-tracking branch 'origin/develop' into develop

bubblobill committed:
Moved some constants to ZoneRendererConstants
Moved utility methods, fogExtents() and  zoneExtents(PlayerView view) to ExportDialog as it was the only place they were used.
Created ZoneCompositor which will take on the role of determining what to draw.
Added ZoneCompositor to ZR constructor
Changed get and restore antialiasing to RenderingHints

bubblobill committed:
Moved some constants to ZoneRendererConstants
Moved utility methods, fogExtents() and  zoneExtents(PlayerView view) to ExportDialog as it was the only place they were used.
Created ZoneCompositor which will take on the role of determining what to draw.
Added ZoneCompositor to ZR constructor
Changed get and restore antialiasing to RenderingHints

bubblobill committed:
Spotlessed

bubblobill committed:
Spotlessed

bubblobill committed:
Moved renderGrid to new class GridRenderer.
Moved drawing of halos to new class HaloRenderer.
moveGridBy() and adjustGridSize() moved to GridRenderer
GridTool updated to moveGridBy's new home.

bubblobill committed:
Moved renderGrid to new class GridRenderer.
Moved drawing of halos to new class HaloRenderer.
moveGridBy() and adjustGridSize() moved to GridRenderer
GridTool updated to moveGridBy's new home.

bubblobill committed:
Merge branch 'RPTools:develop' into develop

bubblobill committed:
Merge branch 'RPTools:develop' into Refactor_ZoneRenderer

cwisniew committed:
Merge branch 'develop' into Refactor_ZoneRenderer

bubblobill committed:
Merge branch 'develop' into Refactor_ZoneRenderer

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/ZoneRenderer.java:
```diff
@@ -85,20 +85,17 @@ public class ZoneRenderer extends JComponent
 
   /** DebounceExecutor for throttling repaint() requests. */
-  private final DebounceExecutor repaintDebouncer;
+  final DebounceExecutor repaintDebouncer;
```
If the suggested changes are made in GridRenderer, we can make this `private` again.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:
```diff
@@ -0,0 +1,75 @@
+    }
+    zone.getGrid().setOffset(gridOffsetX, gridOffsetY);
+    renderer.repaintDebouncer.dispatch();
+  }
```
We can merge these into the GridTool instead, which will avoid the need to expose the grid renderer publicly to the tool. Also there is a `GridChanged` event that will be fired when `Grid.setOffset()` or `Grid.setSize()` is called - `ZoneRenderer` already listens for that in order to repaint itself, so explicit requests to do so shouldn't be necessary.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/ZoneRenderer.java:
```diff
@@ -1034,10 +915,21 @@ private boolean shouldRenderLayer(Layer layer, PlayerView view) {
+    markerLocationList.clear();
+    itemRenderList.clear();
+
+    if (!compositor.isInitialised()) compositor.setRenderer(this);
```
Let's initialize the renderers here as well. Noticing also that the grid renderer is redundantly checked.

bubblobill replied:
Not redundant.
Whilst the zoneRenderer is under construction we cannot pass it as an argument to the other renderers.
This/these ensure that the renderer is set on the renderers on the first run through.
Still going to group them all together at the start as requested.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/ZoneRenderer.java:
```diff
@@ -85,20 +85,17 @@ public class ZoneRenderer extends JComponent
-
-  private static final Color TRANSLUCENT_YELLOW =
-      new Color(Color.yellow.getRed(), Color.yellow.getGreen(), Color.yellow.getBlue(), 50);
+  protected static final ZoneRendererConstants constants = new ZoneRendererConstants();
```
No need to instantiate `ZoneRendererConstants`, it can be used statically where needed.

bubblobill committed:
Merge branch 'develop' into Refactor_ZoneRenderer

bubblobill committed:
Merge remote-tracking branch 'origin/Refactor_ZoneRenderer' into Refactor_ZoneRenderer

bubblobill committed:
Removed instantiation of ZoneRendererConstants
adjustGridSize(int) & moveGridBy(int,int) relocated to GridTool and removed repaintDebouncer call.
Re-Privatised repaint debouncer.
Grouped renderer initialisers together at beginning of renderZone.
Tested GridTool successfully and put a halo on a token so it all seems to work.

bubblobill commented:
Changes made as requested.
Tested successfully.

bubblobill commented:
No idea why it's failing formatting. It's had its spotless apply.

cwisniew commented:
> No idea why it's failing formatting. It's had its spotless apply.

Is your fork up to date and running the version of spotless from the have 21 upgrade?

bubblobill committed:
Merge branch 'RPTools:develop' into Refactor_ZoneRenderer

bubblobill committed:
Discovered halo was rendering with incorrect bounds, was not updating with pan and zoom. Fixed by passing TokenLocation and using its area.
GraphicsUtil and ColorComboBoxRenderer have not been changed but appear to have succumbed to an imperceptible spootless modification.

bubblobill commented:
Is all fixed and happy now

bubblobill committed:
Merge branch 'RPTools:develop' into Refactor_ZoneRenderer

bubblobill committed:
Merge branch 'RPTools:develop' into Refactor_ZoneRenderer
